### PR TITLE
Refactor/improve validations

### PR DESCRIPTION
**Pull request by andersonnaz**
**Testes dos Validations**

**Descrição:**
Foram feitos testes para todos os validadores de entrada de dados.
Resolvido bug onde quebrava o código quando o valor passado nos parâmetros do body não eram do tipo string, quebrava o código retornando o um erro 500.

**Testes Realizados:**
Foram feitos testes em todos os arquivos, garantindo cobertura de 100%.

**Informações Adicionais:**
As Classes PasswordValidation e CompareFieldsValidation recebe um customError que é utilizado para retornar um erro diferente, é realmente necessário deixar aberto para utilizar um erro customizado?

### CODE REVIEW

File: src/application/validation/validators/password-validation.ts [hunk 1]
```diff
@@ -10,6 +10,9 @@ export class PasswordValidation implements Validation {
   ) {}
 
   validate (input: any): Error | void {
+    if (typeof input[this.fieldName] !== 'string') {
+      return new InvalidParamError(this.fieldName)
+    }
     const isValid = this.validator.isValid(input[this.fieldName])
     if (!isValid) {
       if (this.customError) {
```
Comment (rodrigofmeneses) [inline src/application/validation/validators/password-validation.ts:18]: About custom error, I think that this is important to return more descriptive error for caller. On body validation an "InvalidParamError" is too generic, just means that the param is not an string or is not provided. For example, password validation case we need show to caller that params are correctly sended but they missmatch in values.

File: tests/src/application/validation/validators/email-validation.spec.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    sut,
+    emailValidatorStub
+  }
+}
+
+describe('EmailValidation', () => {
+  it('should returns InvalidParamError if fieldName is not a string', () => {
+    const { sut } = makeSut()
+
+    const result = sut.validate({ fieldName: 11 })
```
Comment (rodrigofmeneses) [inline tests/src/application/validation/validators/email-validation.spec.ts]: to be more descriptive you can use an name for this value 11. Like `notStringValue = 11`

File: tests/src/application/validation/validators/email-validation.spec.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    expect(result).toStrictEqual(new InvalidParamError('fieldName'))
+  })
+
+  it('should returns InvalidParamError if fieldName is not a valid email', () => {
+    const { sut, emailValidatorStub } = makeSut()
+    jest.spyOn(emailValidatorStub, 'isValid').mockReturnValueOnce(false)
+
+    const result = sut.validate({ fieldName: 'invalid_email@mail.com' })
```
Comment (rodrigofmeneses) [inline tests/src/application/validation/validators/email-validation.spec.ts]: Same here. Maybe you want avoid hardcoded strings and use variables or values with better descriptions

File: tests/src/application/validation/validators/name-validation.spec.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    sut,
+    nameValidatorStub
+  }
+}
+
+describe('NameValidation', () => {
+  it('should returns invalidParamError if fieldName is not a string', () => {
+    const { sut } = makeSut()
+
+    const result = sut.validate({ fieldName: 11 })
```
Comment (rodrigofmeneses) [inline tests/src/application/validation/validators/name-validation.spec.ts]: Same in this file.
https://github.com/PetJournal/petjournal.api/pull/40#discussion_r1324466939

File: tests/src/application/validation/validators/name-validation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  const nameValidatorStub = makeNameValidatorStub()
+  const sut = new NameValidation(fakeFieldName, nameValidatorStub)
+  return {
+    sut,
+    nameValidatorStub
+  }
+}
+
+describe('NameValidation', () => {
+  it('should returns invalidParamError if fieldName is not a string', () => {
```
Comment (rodrigofmeneses) [inline tests/src/application/validation/validators/name-validation.spec.ts]: I usualy group tests by dependency. For example, `NameValidation` has two dependencies, `fieldName` and `nameValidator`, so I think that is interesting make two groups. You can create this "dependencies groups" with `describe('NameOfGroup', => {...})`. It's not a strict rule, but usualy when we use the Stub, like `nameValidatorStub`, we are in group of `nameValidator`. What you think?

File: tests/src/application/validation/validators/name-validation.spec.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    sut.validate({ fieldName: 'valid_name' })
+
+    expect(isValidSpy).toHaveBeenCalledWith('valid_name')
+  })
+
+  it('should return void if fieldName is a valid name', () => {
+    const { sut } = makeSut()
+
+    const result = sut.validate({ fieldName: 'valid_name' })
```
Comment (rodrigofmeneses) [inline tests/src/application/validation/validators/name-validation.spec.ts]: The object `{ fieldName: 'valid_name' }` it is used many times, maybe you consider create a global object. When you need change the value or be explicit of intention like 'invalid_name', you can use other object or overwrite.

File: tests/src/application/validation/validators/password-validation.spec.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
+import { InvalidParamError } from '@/application/errors'
+import { PasswordValidation, type PasswordValidator } from '@/application/validation'
+
+const makePasswordValidatorStub = (): PasswordValidator => {
```
Comment (rodrigofmeneses) [inline tests/src/application/validation/validators/password-validation.spec.ts]: We have a proper folder to group the Stubs and make test file more clean, specially if you will reutilize, check it

File: tests/src/application/validation/validators/required-field-validation.spec.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  const fakeFieldName: string = 'fieldName'
+  const sut = new RequiredFieldValidation(fakeFieldName)
+  return sut
+}
+
+describe('RequiredFieldValidation', () => {
+  it('should return MissingParamError if fieldName is not provided', () => {
+    const sut = makeSut()
+
+    const result = sut.validate({ fieldName: '' })
```
Comment (rodrigofmeneses) [inline tests/src/application/validation/validators/required-field-validation.spec.ts]: You can test when a empty object is sended too.

**Review by rodrigofmeneses — CHANGES_REQUESTED** [review]
Very good job, consider the refactors to improve organization, but the logic is excelent.

File: tests/src/application/validation/validators/name-validation.spec.ts (hunk 5 above)
Comment (andersonnaz) [reply tests/src/application/validation/validators/name-validation.spec.ts]: sure, it's make sense.

**Review by rodrigofmeneses — APPROVED** [review]
LGFM

**Review by rodrigofmeneses — APPROVED** [review]
LGFM

**Review by matheusgondra — APPROVED** [review]
LGTM

**Review by matheusgondra — APPROVED** [review]
LGTM

**Pull request merged**